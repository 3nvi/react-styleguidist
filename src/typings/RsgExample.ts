export interface Modifiers {
	/** Render static highlighted code instead of a dynamic preview with a code editor */
	static?: boolean;
	/** Renders an example code but hides the code editor */
	noeditor?: boolean;
	/** Adds padding between example child elements */
	padded?: boolean;
	[key: string]: unknown;
}

export interface CodeExample {
	content: string;
	lang?: string;
	settings?: Modifiers;
}
