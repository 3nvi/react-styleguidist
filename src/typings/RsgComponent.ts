import { MethodDescriptor, PropDescriptor, TagProps } from 'react-docgen';
import { ExamplesModule } from './RsgExample';
import { RequireItResult } from './RsgRequireItResult';

export type ExpandMode = 'expand' | 'collapse' | 'hide';

export interface BaseComponent {
	hasExamples?: boolean;
	name?: string;
	slug?: string;
	href?: string;
	filepath?: string;
	pathLine?: string;
	description?: string;
	usageMode?: ExpandMode;
}

export interface Component extends BaseComponent {
	visibleName?: string;
	props?: {
		displayName?: string;
		visibleName?: string;
		description?: string;
		methods?: MethodDescriptor[];
		props?: PropDescriptor[];
		tags?: TagProps;
		content?: ExamplesModule;
	};
	module?: number;
	metadata?: {
		tags?: string[];
	};
}

export interface LoaderComponent extends BaseComponent {
	module: RequireItResult;
	props: RequireItResult;
	metadata: RequireItResult | Record<string, unknown>;
}
