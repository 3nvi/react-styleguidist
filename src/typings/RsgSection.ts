import { RequireItResult } from './RsgRequireItResult';
import { ExamplesModule } from './RsgExample';
import { LoaderComponent, ExpandMode, Component } from './RsgComponent';

export interface BaseSection {
	name?: string;
	slug?: string;
	ignore?: string | string[];
	description?: string;
	exampleMode?: ExpandMode;
	usageMode?: ExpandMode;
	href?: string;
	sectionDepth?: number;
	external?: boolean;
	expand?: boolean;
}

export interface ProcessedSection extends BaseSection {
	exampleMode: ExpandMode;
	visibleName?: string;
	filepath?: string;
	externalLink?: boolean;
	href?: string;
}

/**
 * Section used on the client
 * It is the output of the function `client/utils/processSection`
 */
export interface Section extends ProcessedSection {
	content?: ExamplesModule;
	components?: Component[];
	sections?: Section[];
}

/**
 * Item of the Table Of Contents
 */
export interface TOCItem {
	name?: string;
	visibleName?: string;
	slug?: string;
	href?: string;
	heading?: boolean;
	shouldOpenInNewTab?: boolean;
	selected?: boolean;
	initialOpen?: boolean;
	forcedOpen?: boolean;
	content?: React.ReactNode;
	components?: Component[];
	sections?: Section[];
}

/**
 * Used in the config file and at the early stages of processing
 * in `schema/config.ts`
 */
export interface RawSection extends BaseSection {
	components?: string | string[] | (() => string[]);
	sections?: RawSection[];
	content?: string;
}

/**
 * Type returned when sections are transformed to their webpack
 * loadable equivalents
 */
export interface LoaderSection extends BaseSection {
	slug?: string;
	content?: RequireItResult;
	components: LoaderComponent[];
	sections: LoaderSection[];
	exampleMode: ExpandMode;
}
