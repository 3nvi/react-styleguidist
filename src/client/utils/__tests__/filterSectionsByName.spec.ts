import deepfreeze from 'deepfreeze';
import filterSectionsByName from '../filterSectionsByName';
import * as Rsg from '../../../typings';

const sectionsRaw: Rsg.Section[] = [
	{
		exampleMode: 'collapse',
		name: 'General',
		sections: [
			{
				exampleMode: 'collapse',
				name: 'Particles',
				components: [
					{
						name: 'Button',
					},
					{
						name: 'Image',
					},
				],
			},
		],
	},
];
const sections = deepfreeze(sectionsRaw);

describe('filterSectionsByName', () => {
	it('should recursively filter sections and components by name', () => {
		const result = filterSectionsByName(sections, 'button');
		expect(result).toMatchSnapshot();
	});

	it('should skip sections without matches inside', () => {
		const result = filterSectionsByName(sections, 'general');
		expect(result).toMatchSnapshot();
	});

	it('should return empty array if no components of sections match query', () => {
		const result = filterSectionsByName(sections, 'pizza');
		expect(result).toEqual([]);
	});
});
