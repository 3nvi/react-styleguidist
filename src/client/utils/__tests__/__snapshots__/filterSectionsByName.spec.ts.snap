// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterSectionsByName should recursively filter sections and components by name 1`] = `
Array [
  Object {
    "components": Array [],
    "exampleMode": "collapse",
    "name": "General",
    "sections": Array [
      Object {
        "components": Array [
          Object {
            "name": "Button",
          },
        ],
        "exampleMode": "collapse",
        "name": "Particles",
        "sections": Array [],
      },
    ],
  },
]
`;

exports[`filterSectionsByName should skip sections without matches inside 1`] = `
Array [
  Object {
    "components": Array [],
    "exampleMode": "collapse",
    "name": "General",
    "sections": Array [],
  },
]
`;
